### DISCUSSION TITLE
Run executable file by pyinstaller generated, the erroe showed “ModuleNotFoundError: No module named 'Notebook'”

### DISCUSSION THREAD

**Discussion by 1893945** [pyinstaller/pyinstaller#7082] [PyInstaller]
Hi, when I Run executable file by pyinstaller generated, the erroe showed “ModuleNotFoundError: No module named 'Notebook'”. 
error tip:
```
[<class 'ModuleNotFoundError'>] No module named 'Notebook'

Traceback (most recent call last):
  File "pygubu\builder.py", line 193, in _import_class
  File "importlib\__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 961, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'ttk'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "pygubu\builder.py", line 201, in _import_class
  File "importlib\__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'ttk'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tktools.py", line 235, in <module>
  File "tktools.py", line 28, in __init__
  File "pygubu\builder.py", line 167, in get_object
  File "pygubu\builder.py", line 215, in _realize
  File "pygubu\builder.py", line 204, in _import_class
  File "importlib\__init__.py", line 127, in import_module
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'Notebook'

 File "tktools.py", line 28 code is:
self.mainwindow = builder.get_object("notebook1", master)
```
so, How should I deal with this problem?

thanks

**Comment by rokm** [comment]
Can you provide a minimum example that triggers this error? The missing `ttk` error might imply that `tkinter.ttk` is not collected, but I cannot reproduce this with basic `pygubu` example.

**Comment by 1893945** [comment]
Hi rokm,
this is the minimum example.
see the attach files
[tk_tools.zip](https://github.com/pyinstaller/pyinstaller/files/9556940/tk_tools.zip)

pyinstaller --add-data "tools2.ui;." -D tktools2.py

pyinstaller               5.4.1
pyinstaller-hooks-contrib 2022.10

note: 
I debugged and found that it seemed to be PluginManager.py "for _, name, _ in iter_namespace(pygubu.plugins):" not executed, when run exe file. 

thanks

**Reply by rokm** [reply]
Hmm, that example seems to work for me in frozen form.

>  I debugged and found that it seemed to be PluginManager.py "for _, name, _ in iter_namespace(pygubu.plugins):" not executed, when run exe file.

Perhaps `pygubu.plugins` are not collected? Can you check `build/tktool2/PYZ-00.toc` and look for entries starting with `pygubu.plugins.`?

`pygubu` provides a PyInstaller [hook](https://github.com/alejandroautalan/pygubu/blob/master/src/pygubu/__pyinstaller/hook-pygubu.py) that ensures that (some of?) the plugins are collected. When you are building the example, do you see line such as this in the console output?

```
8578 INFO: Loading module hook 'hook-pygubu.py' from '[...\\site-packages\\pygubu\\__pyinstaller'...
``` 

But I guess you can also just try to force collection of all those plugins yourself, by adding `--collect-submodules pygubu.plugins`  or even `--collect-submodules pygubu` to your PyInstaller command.

**Reply by 1893945** [reply]
I checked PYZ-00.toc, it have pygubu.plugins only, not have pygubu.plugins.?

![image](https://user-images.githubusercontent.com/7885661/190033832-93f4882c-9d5f-48f5-b71f-a3a166ee3a46.png)

besides, I checked pygubu code, find it have pyinstall hook:
![image](https://user-images.githubusercontent.com/7885661/190034277-636e4589-036c-4063-b833-a06ce0863da2.png)
![image](https://user-images.githubusercontent.com/7885661/190034369-826eba9b-190f-4aef-8900-bc689642562d.png)
![image](https://user-images.githubusercontent.com/7885661/190034405-7c9ffc0f-006c-47fb-9d3b-62551fd6e2fc.png)

8578 INFO: Loading module hook 'hook-pygubu.py' from '[...\\site-packages\\pygubu\\__pyinstaller'...

417 INFO: Platform: Windows-7-6.1.7601-SP1
418 INFO: wrote F:\TK\tktools2\src\tk_tools\tktools.spec
431 INFO: UPX is not available.
433 INFO: Extending PYTHONPATH with paths
['F:\\TK\\tktools2\\src\\tk_tools']
416 WARNING: discover_hook_directories: Failed to process hook entry point 'hook-dirs = pygubu.__pyinstaller:get_hook_dirs': TypeError: 'type' object is not subscriptable
1197 INFO: checking Analysis
1322 INFO: Building because f:\tk\tktools2\venv\lib\site-packages\pygubu\__init__.py changed
1323 INFO: Initializing module dependency graph...
1325 INFO: Caching module graph hooks...
1335 WARNING: Several hooks defined for module 'numpy'. Please take care they do not conflict.
1343 INFO: Analyzing base_library.zip ...
3711 INFO: Loading module hook 'hook-encodings.py' from 'f:\\tk\\tktools2\\venv\\lib\\site-packages\\PyInstaller\\hooks'...

**Reply by rokm** [reply]
> 416 WARNING: discover_hook_directories: Failed to process hook entry point 'hook-dirs = pygubu.__pyinstaller:get_hook_dirs': TypeError: 'type' object is not subscriptable 

So while you do have the pygubu hook, it is not ran because it cannot be loaded due to that type error. What version of python are you using?

Try collecting the plugins without the hook, i..e, use `--collect-submodules pygubu.plugins` or even `--collect-submodules pygubu`.

**Reply by 1893945** [reply]
python 3.8

I put pygubu.plugins.* into spec file(hiddenimports) and build, the executable file can work

Thank you very much for your help

**Reply by 1893945** [reply]
python 3.8

I put pygubu.plugins.* into spec file(hiddenimports) and build, the executable file can work

Thank you very much for your help

**Reply by 1893945** [reply]
this is a bug，pygubu hook in python 3.8.
and pygubu has fixed [Fix pyinstaller hook for python 3.8 refs](https://github.com/alejandroautalan/pygubu/commit/2b9d251e2074ec0a0c78e825743bc74762fe6372)

thank you again
